controllers/topolvm_controller.go

```diff
@@ -0,0 +1,325 @@
+
+func (c topolvmController) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+
+	// get the desired state of topolvm controller deployment
+	controllerDeployment := getControllerDeployment(lvmCluster)
+
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, controllerDeployment, func() error { return nil })
+
+	// log the result based on current reconcile
+	switch result {
+	case cutil.OperationResultCreated:
+		r.Log.Info("csi controller", "operation", result, "name", controllerDeployment.Name)
+	case cutil.OperationResultUpdated:
+		r.Log.Info("csi controller", "operation", result, "name", controllerDeployment.Name)
+	case cutil.OperationResultNone:
+		r.Log.Info("csi controller", "operation", result, "name", controllerDeployment.Name)
+	default:
+		r.Log.Error(err, "csi controller reconcile failure", "name", controllerDeployment.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (c topolvmController) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	controllerDeployment := &appsv1.Deployment{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmControllerDeploymentName, Namespace: lvmCluster.Namespace}, controllerDeployment)
+
+	if err != nil {
+		// already deleted in previous reconcile
+		if errors.IsNotFound(err) {
+			r.Log.Info("csi controller deleted", "TopolvmController", controllerDeployment.Name)
+			return nil
+		}
+		r.Log.Error(err, "failed to retrieve csi controller deployment", "TopolvmController", controllerDeployment.Name)
+		return err
+	}
+
+	// if not deleted, initiate deletion
+	if controllerDeployment.GetDeletionTimestamp().IsZero() {
+		if err = r.Client.Delete(ctx, controllerDeployment); err != nil {
+			r.Log.Error(err, "failed to delete topolvm controller deployment", "TopolvmController", controllerDeployment.Name)
+			return err
+		}
+	} else {
+		// set deletion in-progress for next reconcile to confirm deletion
+		return fmt.Errorf("topolvm controller deployment %s is already marked for deletion", controllerDeployment.Name)
+	}
+
+	return nil
+}
+
+func (c topolvmController) updateStatus(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	// TODO: Verify the status of controller plugin deployment and set the same on CR
+	return nil
+}
+
+func getControllerDeployment(lvmCluster *lvmv1alpha1.LVMCluster) *appsv1.Deployment {
+
+	// Topolvm CSI Controller Deployment
+	var replicas int32 = 1
+	volumes := []corev1.Volume{
+		{Name: "socket-dir", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
+		{Name: "certs", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
+	}
+
+	// TODO: Remove custom generation of TLS certs, current it's being used in topolvm controller manager
+	initContainers := []corev1.Container{
+		*getInitContainer(),
+	}
+
+	// getall containers that are part of csi controller deployment
```

Comment:
```suggestion
	// get all containers that are part of csi controller deployment
```

Reply:
ack, can I take this as part of separate PR?

Reply:
ack, can I take this as part of separate PR?

Reply:
sure. Separate PR sounds good.

Reply:
done